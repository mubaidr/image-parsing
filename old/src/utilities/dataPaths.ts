import os from 'os'
import path from 'path'

const isDev: boolean = process.env.NODE_ENV === 'development'
const tmp: string = isDev ? path.resolve('.tmp') : os.tmpdir()
const home: string = path.join(os.homedir(), 'desktop')

const dataPaths = {
  design: path.resolve('__tests__', '_data', 'design.qr.svg'),
  designBarcode: path.resolve('__tests__', '_data', 'design.svg'),
  home,
  images: path.resolve('__tests__', '_data', 'images-qrcode'),
  imagesBarcode: path.resolve('__tests__', '_data', 'images-barcode'),
  key: path.resolve('__tests__', '_data', 'key.xlsx'),
  keyImage: path.resolve('__tests__', '_data', 'key.jpg'),
  questionsModel: isDev
    ? path.resolve('src/data/questions-model.json')
    : path.resolve('dist/data/questions-model.json'),
  result: path.resolve('__tests__', '_data', 'result.xlsx'),
  resultCompiled: path.resolve('__tests__', '_data', 'resultCompiled.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
}

export { dataPaths }
