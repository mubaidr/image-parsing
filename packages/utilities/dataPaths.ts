import os from 'os';
import path from 'path';

const tmp = os.tmpdir();

export const DataPaths = {
  design: path.resolve('tests', '_data', 'design.qr.svg'),
  designBarcode: path.resolve('tests', '_data', 'design.svg'),
  home: os.homedir(),
  images: path.resolve('tests', '_data', 'images-qrcode'),
  imagesBarcode: path.resolve('tests', '_data', 'images-barcode'),
  key: path.resolve('tests', '_data', 'key.xlsx'),
  keyImage: path.resolve('tests', '_data', 'key.jpg'),
  result: path.resolve('tests', '_data', 'result.xlsx'),
  resultCompiled: path.resolve('tests', '_data', 'resultCompiled.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
};
