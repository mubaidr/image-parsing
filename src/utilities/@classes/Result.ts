import uuid from 'uuid'
import IAnswer from '../@interfaces/IAnswer'

class Result {
  private id: string
  public list: IAnswer = {}
  public rollNo: string = ''
  public imageFile: string = ''

  public constructor(rollNo: string, imageFile: string) {
    this.id = uuid()
    this.rollNo = rollNo
    this.imageFile = imageFile
  }

  public addAnswer(title: string, value: string) {
    this.list[title] = { value }
  }

  public isKey(): boolean {
    return this.rollNo === 'key'
  }

  public hasValidRollNo(): boolean {
    return !this.isKey() && this.rollNo !== ''
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public static fromJson(obj: any): Result {
    throw 'Not Implemented'
  }
}

export default Result
