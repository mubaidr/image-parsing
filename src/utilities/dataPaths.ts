import os from 'os'
import path from 'path'

const isDev: boolean = process.env.NODE_ENV === 'development'

const tmp: string = isDev ? path.resolve('.tmp') : os.tmpdir()
const home: string = path.join(os.homedir(), 'desktop')

const dataPaths = {
  design: path.resolve('_test_data', 'design.qr.svg'),
  designBarcode: path.resolve('_test_data', 'design.svg'),
  home,
  images: path.resolve('_test_data', 'images-qrcode'),
  imagesBarcode: path.resolve('_test_data', 'images-barcode'),
  key: path.resolve('_test_data', 'key.xlsx'),
  keyImage: path.resolve('_test_data', 'key.jpg'),
  questionsModel: path.resolve('src/data/questions-model.json'),
  result: path.resolve('_test_data', 'result.xlsx'),
  root: path.resolve('.'),
  temp: tmp,
  tmp,
}

export { dataPaths }
