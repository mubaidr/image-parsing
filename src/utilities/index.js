const brain = require('brain.js')
const cheerio = require('cheerio')
const childProcess = require('child_process')
const fastGlob = require('fast-glob')
const fs = require('fs')
const javascriptBarcodeReader = require('javascript-barcode-reader')
const os = require('os')
const download = require('downloadjs')
const NO_OF_CORES = require('physical-cpu-count')
const dataPaths = require('./data-paths')
// const javascriptBarcodeReader = require('../../../Javascript-Barcode-Reader/src')

/**
 * Create worker process equal to cpu cores
 *
 * @returns {Array} array of child process forks
 */
async function createWorkerProcesses(imagesCount) {
  const WORKERS = []

  const availMemory = (os.totalmem() - os.freemem()) / (1024 * 1024 * 1024)
  let CORE_COUNT = Math.min(NO_OF_CORES, imagesCount)

  // If available ram is less than 200MB/400MB, use only one/two worker processes respectively
  if (availMemory < 0.2) {
    CORE_COUNT = 1
  } else if (availMemory < 0.4) {
    CORE_COUNT = 2
  }

  for (let i = 0; i < CORE_COUNT; i += 1) {
    WORKERS.push(
      childProcess.fork(`${__dirname}/processTaskWorker.js`, [], {
        silent: true,
      }),
    )
  }

  return WORKERS
}

/**
 * Extracts position & dimensions of objects from SVG design File
 *
 * @param {String} designFilePath Path to the svg design file
 * @returns {Object} JSON object
 */
async function getDesignData(dir) {
  const designData = {
    questions: {},
  }
  const ROLL_NO_PATTERN = new RegExp(/rollnobarcode/gi)
  const QUESTION_PATTERN = new RegExp(/(q[1-9][0-9]?[ad])\b/gi)

  const $ = cheerio.load(fs.readFileSync(dir, 'utf8'))
  const svgViewBox = $('svg')[0].attribs.viewBox.split(' ')

  designData.width = Math.ceil(svgViewBox[2] - svgViewBox[0])
  designData.height = Math.ceil(svgViewBox[3] - svgViewBox[1])

  let gotRollNo = false
  let x
  let y
  let rx
  let ry
  let width
  let height
  let left
  let top

  const groups = $('g')
  for (let i = 0; i < groups.length; i += 1) {
    const group = groups[i]

    const title = $(group)
      .find('title')
      .first()
      .html()
      .trim()
      .toLowerCase()

    const isQuestionGroup = QUESTION_PATTERN.test(title)
    const isRollNoGroup =
      isQuestionGroup || gotRollNo ? false : ROLL_NO_PATTERN.test(title)

    if (isQuestionGroup || isRollNoGroup) {
      const transform = $(group)
        .attr('transform')
        .replace(/(translate)|\(|\)/gi, '')
        .split(',')
        .map(val => parseFloat(val))

      const rect = $(group)
        .find('rect')
        .first()

      left = parseFloat(rect.attr('x'))
      top = parseFloat(rect.attr('y'))

      rx = parseFloat(rect.attr('rx')) || 0
      ry = parseFloat(rect.attr('ry')) || 0

      x = left - rx + transform[0]
      y = top - ry + transform[1]

      width = parseFloat(rect.attr('width')) + rx
      height = parseFloat(rect.attr('height')) + ry
    }

    if (isQuestionGroup) {
      const optionTitle = title.slice(-1)
      const questionNumber = title.slice(0, -1)

      if (!designData.questions[questionNumber]) {
        designData.questions[questionNumber] = {}
      }

      if (optionTitle === 'a') {
        designData.questions[questionNumber].x1 = x
        designData.questions[questionNumber].y1 = y
      } else {
        designData.questions[questionNumber].x2 = x + width
        designData.questions[questionNumber].y2 = y + height
      }
    }

    if (isRollNoGroup) {
      designData.rollNo = { x1: x, y1: y, x2: x + width, y2: y + height }
      gotRollNo = true
    }
  }

  return designData
}

/**
 * Return a list of valid image format files from the provided path
 *
 * @param {String} path Path to sarch for images
 * @param {Array.<String>} format Array of extensions of valid image formats
 * @returns {Array.<String>} List of file paths
 */
function getImagePaths(dir) {
  const formats = [
    'png',
    'jpg',
    'jpeg',
    'jpe',
    // 'jfif',
    'gif',
    'tif',
    'tiff',
    'bmp',
    // 'dib',
  ]

  return fastGlob(`${dir}/*.{${formats.join(',')}}`, { onlyFiles: true })
}

/**
 * Returns a trained neural network function
 *
 * @returns {Function} Neural network function
 */
function getNeuralNet(src) {
  const net = new brain.NeuralNetwork()

  return net.fromJSON(
    JSON.parse(fs.readFileSync(src || dataPaths.trainingData)),
  )
}

/**
 *
 *
 * @param {Object} designData A JSON Object containing information about the position, width, height of elements in svg design file (available from utiltities/getDesignData)
 * @param {String} path Path of scanned image file
 * @param {Object=} resultsData Path to csv file for training data
 * @param {Number=} rollNo Roll no of the current scanned image
 * @returns {Object} {title: {String}, data: {buffer}}
 */
async function getQuestionsData(designData, img, resultsData, rollNo) {
  const SCALE = 0.25
  const IS_TEST_DATA = resultsData && rollNo

  return new Promise((resolveCol, rejectCol) => {
    img.resize(designData.width * SCALE).max()

    const promises = []
    // extract all questions portions
    Object.keys(designData.questions).forEach(title => {
      const p = new Promise(resolve => {
        const q = designData.questions[title]

        img
          .extract({
            left: Math.floor(q.x1 * SCALE),
            top: Math.floor(q.y1 * SCALE),
            width: Math.ceil((q.x2 - q.x1) * SCALE),
            height: Math.ceil((q.y2 - q.y1) * SCALE),
          })
          // .png()
          // .toFile(path.join(dataPaths.tmp, `${rollNo}-${title}.png`), err => {
          //   if (err) console.log(err)
          //   resolve()
          // })
          .toBuffer({ resolveWithObject: true })
          .then(({ data }) => {
            const binaryData = []

            for (let i = 0; i < data.length; i += 3) {
              const r = i
              const g = i + 1
              const b = i + 2
              const avg = Math.ceil((data[r] + data[g] + data[b]) / 3)
              const threshold = 15

              if (avg <= 80) {
                // black pixel
                binaryData.push(0)
              } else if (
                data[r] <= avg + threshold &&
                data[r] >= avg - threshold &&
                (data[g] <= avg + threshold && data[g] >= avg - threshold) &&
                (data[b] <= avg + threshold && data[b] >= avg - threshold)
              ) {
                // grey pixel
                binaryData.push(1)
              } else {
                // color pixel
                binaryData.push(0)
              }
            }

            if (IS_TEST_DATA) {
              // for training data
              if (resultsData[rollNo] && resultsData[rollNo][title] !== '*') {
                const o = {}
                o[resultsData[rollNo][title]] = 1

                resolve({ input: binaryData, output: o })
              } else {
                resolve(false)
              }
            } else {
              // for processing data
              resolve({ title, data: binaryData })
            }
          })
      })

      promises.push(p)
    })

    Promise.all(promises)
      .then(res => {
        resolveCol(res)
      })
      .catch(err => {
        rejectCol(err)
      })
  })
}

/**
 *
 *
 * @param {Object} designData A JSON Object containing information about the position, width, height of elements in svg design file (available from utiltities/getDesignData)
 * @param {String} path Path of scanned image file
 * @returns {String} Roll Number
 */
async function getRollNoFromImage(designData, img) {
  const metadata = await img.metadata()
  const rollNoPos = designData.rollNo
  const ratio = metadata.width / designData.width
  const width = Math.ceil((rollNoPos.x2 - rollNoPos.x1) * ratio)
  const height = Math.ceil((rollNoPos.y2 - rollNoPos.y1) * ratio)

  const { data, info } = await img
    .extract({
      left: Math.floor(rollNoPos.x1 * ratio),
      top: Math.floor(rollNoPos.y1 * ratio),
      width,
      height,
    })
    .toBuffer({ resolveWithObject: true })

  const newData = []
  if (info.channels < 4) {
    for (let i = 0; i < data.length; i += info.channels) {
      // copy current channels
      for (let j = 0; j < info.channels; j += 1) {
        newData.push(data[i + j])
      }
      // add missing channels
      for (let k = 0; k < 4 - info.channels; k += 1) {
        newData.push(255)
      }
    }
  }

  return javascriptBarcodeReader(
    { data: info.channels < 4 ? newData : data, width, height },
    { barcode: 'code-39' },
  )
}

/**
 *
 *
 * @param {String} str JSON object or JSON file path
 * @param {Boolean} isPath True for path
 * @param {Boolean} isKey True if the CSV file is answer key file
 * @returns {Object} CSV String
 */
async function CSVToJSON(str, isPath, isKey) {
  const resultFile = isPath ? fs.readFileSync(str, 'utf8') : str
  const json = {}

  // extract rows and header names
  const rows = resultFile.split('\n')
  const headerValues = rows[0]
    .replace(/ /gi, '')
    .toLowerCase()
    .split(',')

  // extract index of roll no in header row
  const rollNoIndex = Math.max(
    headerValues.indexOf('rollno'),
    headerValues.indexOf('roll#'),
    headerValues.indexOf('rollnumber'),
  )

  if (isKey) rows.length = 2
  // iterate rows but skips header row
  for (let i = 1; i < rows.length; i += 1) {
    const obj = {}
    const values = rows[i]
      .replace(/(\s)|(\.)|(-)|(_)/gi, '')
      // .toLowerCase()
      .split(',')

    values.forEach((value, index) => {
      if (index !== rollNoIndex) {
        obj[headerValues[index]] = value
      }
    })

    if (isKey) {
      json.key = obj
    } else {
      const roll = values[rollNoIndex]

      if (roll) {
        json[roll] = obj
      }
    }
  }

  return json
}

/**
 *
 *
 * @param {String} str JSON object or JSON file path
 * @param {Boolean} isPath True for path
 * @returns {Object} CSV String
 */
function JSONToCSV(str, isPath) {
  const obj = isPath ? JSON.parse(fs.readFileSync(str, 'utf8')) : str

  let header = 'ROLL NO'
  let csv = ''

  // sort by roll number
  obj.sort((a, b) => Object.keys(a)[0] - Object.keys(b)[0])

  // parse to csv
  Object.values(obj).forEach((val, index) => {
    const [[rollno, keys]] = Object.entries(val)

    // eslint-disable-next-line
    csv += '\n'
    csv += rollno

    // sort by question no
    const keyEntries = Object.entries(keys).sort(
      (a, b) => a[0].replace('q', '') - b[0].replace('q', ''),
    )

    // append to csv
    for (let i = 0; i < keyEntries.length; i += 1) {
      if (i !== keyEntries.length) {
        csv += ','
      }

      csv += keyEntries[i][1].toUpperCase
        ? keyEntries[i][1].toUpperCase()
        : keyEntries[i][1]

      if (index === 0) {
        header += `,${keyEntries[i][0].toUpperCase()}`
      }
    }
  })

  return `${header}${csv}`
}

/**
 * Exports result data
 *
 * @param {Object} resultData Result data JSON format
 * @param {String} name Name of the output file
 * @param {Boolean} isCSV if the provided data is CSV
 */
async function exportResult(resultData, name, isCSV) {
  const csv = isCSV ? resultData : await JSONToCSV(resultData)

  try {
    download(csv, name ? `${name}.csv` : 'output.csv', 'text/csv')
  } catch (error) {
    console.log('output: ', csv)
  }
}

/**
 * Compare the provided result with key to generate marks
 *
 * @param {Object} keys JSON Object containing answer keys
 * @param {Object} result JSON Object containing result keys
 * @param {Number} correctMarks Number of marks to assign per correct answer
 * @param {Number} negativeMarks Number of marks to deduct per correct answer
 * @returns {Number} Total marks obtained
 */
async function compileResult(keys, results, correctMarks, negativeMarks) {
  const keyEntries = Object.entries(keys.key)
  const output = []
  let totalMarks = 0

  for (let i = 0; i < keyEntries.length; i += 1) {
    const [, value] = keyEntries[i]

    if (value === '?' || value === '*') continue

    totalMarks += correctMarks
  }

  Object.entries(results).forEach(([rollNo, answers]) => {
    let marks = 0

    for (let i = 0; i < keyEntries.length; i += 1) {
      const [key, value] = keyEntries[i]

      // if user has not selected any option
      if (value === '?') continue
      // if question has multiple correct options
      if (value === '*') continue
      // if question has no right option
      if (answers[key] === '?') continue

      if (answers[key] === value) {
        // correct option
        marks += correctMarks
      } else {
        // in-correct option
        marks -= negativeMarks
      }
    }

    output.push({
      [rollNo]: {
        marks,
        totalMarks,
      },
    })
  })

  // export
  exportResult(await JSONToCSV(output), 'ResultMarks', true)
}

module.exports = {
  createWorkerProcesses,
  getDesignData,
  getImagePaths,
  getNeuralNet,
  getRollNoFromImage,
  CSVToJSON,
  JSONToCSV,
  getQuestionsData,
  exportResult,
  compileResult,
}
