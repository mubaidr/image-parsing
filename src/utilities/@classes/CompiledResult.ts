import uuid from 'uuid'

import { importExcelToJson } from '../excel'
import Result from './Result'

class CompiledResult {
  private id: string
  private lastSavedTime: Date | undefined

  private keys: Result[] = []
  private results: Result[] = []

  public constructor() {
    this.id = uuid()
  }

  public getKeyCount(): number {
    return this.keys.length
  }

  public getResultCount(): number {
    return this.results.length
  }

  public getKeys(): Result[] {
    return this.keys
  }

  public getResults(): Result[] {
    return this.results
  }

  public addKeys(key: Result | Result[]): CompiledResult {
    if (key instanceof Result) {
      this.keys.push(key)
    } else {
      this.keys.push(...key)
    }

    return this
  }

  public addResults(result: Result | Result[]): CompiledResult {
    if (result instanceof Result) {
      this.results.push(result)
    } else {
      this.results.push(...result)
    }

    return this
  }

  public compile(marks?: number, negativeMarks?: number): CompiledResult {
    if (this.keys.length === 0) throw 'Keys not loaded'
    if (this.results.length === 0) throw 'Results not loaded'

    for (let i = 0; i < this.keys.length; i += 1) {
      for (let j = 0; j < this.results.length; j += 1) {
        this.results[j].compile(this.keys[i], marks, negativeMarks)
      }
    }

    return this
  }

  public static loadFromExcel(src: string): CompiledResult {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const rows: any[] = importExcelToJson(src)
    const compiledResult = new CompiledResult()

    rows.forEach(row => {
      const result = Result.fromJson(row)

      if (result.isKey()) {
        compiledResult.addKeys(result)
      } else {
        compiledResult.addResults(result)
      }
    })

    return compiledResult
  }

  public addFromExcel(src: string): CompiledResult {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const rows: any[] = importExcelToJson(src)

    rows.forEach(row => {
      const result = Result.fromJson(row)

      if (result.isKey()) {
        this.addKeys(result)
      } else {
        this.addResults(result)
      }
    })

    return this
  }

  public static merge(compiledResults: CompiledResult[]): CompiledResult {
    const compiledResult = new CompiledResult()

    compiledResults.forEach(cr => {
      compiledResult.addKeys(cr.getKeys()).addResults(cr.getResults())
    })

    return compiledResult
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public export(): any[] {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const obj: any[] = []

    this.results.forEach(result => {
      obj.push(result.toJson())
    })

    return obj
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public static export(compiledResult: CompiledResult): any[] {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const obj: any[] = []

    compiledResult.results.forEach(result => {
      obj.push(result.toJson())
    })

    return obj
  }
}

export default CompiledResult
