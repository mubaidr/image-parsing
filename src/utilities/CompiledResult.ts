import { v4 as uuidv4 } from 'uuid'
import { importExcelToJson } from './excel'
import { Result } from './Result'

export class CompiledResult {
  id: string
  lastSavedTime: Date | undefined
  keys: Result[] = []
  results: Result[] = []

  constructor() {
    this.id = uuidv4()
  }

  static loadFromExcel(src: string): CompiledResult {
    const rows: unknown[] = importExcelToJson(src)
    const compiledResult = new CompiledResult()

    rows.forEach((row) => {
      const result = Result.fromJson(row)

      if (result.isKey()) {
        compiledResult.addKeys([result])
      } else {
        compiledResult.addResults([result])
      }
    })

    return compiledResult
  }

  addKeys(keys: Result[]): CompiledResult {
    this.keys.push(...keys)
    return this
  }

  addResults(results: Result[]): CompiledResult {
    this.results.push(...results)
    return this
  }

  getKeysAndResults(): Result[] {
    return [...this.keys, ...this.results]
  }

  compile(marks?: number, negativeMarks?: number): CompiledResult {
    if (this.keys.length === 0) throw new Error('Keys not loaded')
    if (this.results.length === 0) throw new Error('Results not loaded')

    for (let i = 0; i < this.keys.length; i += 1) {
      for (let j = 0; j < this.results.length; j += 1) {
        this.results[j].compile(this.keys[i], marks, negativeMarks)
      }
    }

    return this
  }

  sortResults(): CompiledResult {
    this.results.sort((a, b) => {
      return parseInt(a.rollNo || '0', 10) - parseInt(b.rollNo || '0', 10)
    })

    return this
  }

  static merge(compiledResults: CompiledResult[]): CompiledResult {
    const compiledResult = new CompiledResult()

    compiledResults.forEach((cr) => {
      compiledResult.addKeys(cr.keys).addResults(cr.results)
    })

    return compiledResult
  }
}
