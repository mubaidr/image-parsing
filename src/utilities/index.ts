import brain, { INeuralNetworkJSON } from 'brain.js'
import fs from 'fs'

import Result from './@classes/Result'
import KeyNativeEnum from './@enums/KeyNativeEnum'
import { dataPaths } from './dataPaths'
import { getDesignData } from './design'
import { importExcelToJson } from './excel'
import extractTask from './workers/workerExtract'

const getQuestionsNeuralNet = (): brain.NeuralNetwork => {
  const text = fs.readFileSync(dataPaths.questionsModel).toString()
  const json: INeuralNetworkJSON = JSON.parse(text)

  return new brain.NeuralNetwork().fromJSON(json)
}

const convertToBitArray = (data: number[], channels: number): number[] => {
  const binaryData: number[] = []

  for (let i = 0, dataLength = data.length; i < dataLength; i += channels) {
    const threshold = 15
    const thresholdBlack = 80
    const [r, g, b] = data.slice(i, i + channels)
    const avg = Math.ceil((r + g + b) / channels)
    const upperLimit = avg + threshold
    const lowerLimit = avg - threshold

    if (avg <= thresholdBlack) {
      // Black pixel
      binaryData.push(0)
    } else if (
      r <= upperLimit &&
      r >= lowerLimit &&
      g <= upperLimit &&
      g >= lowerLimit &&
      b <= upperLimit &&
      b >= lowerLimit
    ) {
      // Grey pixel
      binaryData.push(1)
    } else {
      // Color pixel
      binaryData.push(0)
    }
  }

  return binaryData
}

const readKey = async (src: string): Promise<Result[] | undefined> => {
  const ext = src.split('.').pop()

  if (ext === undefined) throw new Error('Invalid path specified')

  if (ext in KeyNativeEnum) {
    const rows = importExcelToJson(src)
    const results: Result[] = []

    for (let i = 0, len = rows.length; i < len; i += 1) {
      results.push(Result.fromJson(rows[i]))
    }

    return results
  }

  console.log('yehn group is here.... !!!!')

  const designData = getDesignData(dataPaths.designBarcode)
  const keys = await extractTask.start(
    {
      designData,
      imagePaths: [src],
    },
    false,
  )

  return keys
}

export { convertToBitArray, getQuestionsNeuralNet, readKey }
