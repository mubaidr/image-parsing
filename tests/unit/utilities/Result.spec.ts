import { CompiledResult } from '@/utilities/CompiledResult'
import { dataPaths } from '@/utilities/dataPaths'
import { readKey } from '@/utilities/readKey'
import { Result } from '@/utilities/Result'
import path from 'path'

describe('Result', () => {
  test('should have defaults set', async () => {
    const rollNo = '10023'
    const imageFile = path.resolve(dataPaths.imagesBarcode, '10023.jpg')
    const result = new Result(rollNo, imageFile)

    expect(result.id).toBeDefined()
    expect(result.rollNo).toBe(rollNo)
    expect(result.filePath).toBe(imageFile)
  })

  test('should load from json', async () => {
    const o = {
      correctCount: 0,
      incorrectCount: 0,
      isCompiled: false,
      marks: 0,
      skippedCount: 0,
      totalMarks: 0,
      unattemptedCount: 0,
      id: '183b5aad-7122-4947-99e9-75c2f73cb076',
      imageFile:
        'D:\\Current\\image-parsing\\__tests__\\_data\\images-barcode\\10023.jpg',
      isRollNoExtracted: true,
      post: '',
      questionPaperType: '',
      rollNo: '10023',
      testCenter: '',
      testTime: '',
      q1: 'a',
      q2: 'b',
      q3: 'c',
      q4: 'd',
      q5: '?',
      q6: '*',
    }

    const result = Result.fromJson(o)

    expect(result).toMatchSnapshot()
  })

  test('should be able to export to json object', async () => {
    const compiledResult = CompiledResult.loadFromExcel(dataPaths.result)
    const result = compiledResult.results.reverse()[0]

    expect(result.hasValidRollNo()).toBeTruthy()
    expect(result.hasImageFile()).toBeTruthy()

    const o = result.toJson()

    expect(o).toMatchSnapshot()
  })

  test('should compile', async () => {
    const compiledResult = CompiledResult.loadFromExcel(dataPaths.result)
    const result = compiledResult.results[0]
    const keys = await readKey(dataPaths.key)

    if (!keys) fail()

    keys.forEach((key) => {
      result.compile(key, 3, 1)
    })

    expect(result.totalMarks).toBeGreaterThan(0)

    const isCompiled =
      result.correctCount ||
      result.incorrectCount ||
      result.unattemptedCount ||
      result.skippedCount

    expect(isCompiled).toBeGreaterThan(0)
  })

  test('should be able to identify key and result', async () => {
    const compiledResult = CompiledResult.loadFromExcel(dataPaths.result)
    const result = compiledResult.results[0]
    const keys = await readKey(dataPaths.key)

    expect(result.isKey()).toBeFalsy()

    if (keys) {
      const key = keys[0]
      expect(key.isKey()).toBeTruthy()
    }
  })

  test('should be able to calculate marks for compiled result', async () => {
    const compiledResult = CompiledResult.loadFromExcel(dataPaths.result)
    const result = compiledResult.results[0]
    const keys = await readKey(dataPaths.key)

    if (keys) {
      result.compile(keys[0]).setMarks(3, 1)
      expect(result.totalMarks).not.toBe(0)
    }
  })
})
