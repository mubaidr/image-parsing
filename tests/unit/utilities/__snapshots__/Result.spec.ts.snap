// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result should be able to export to json object 1`] = `
Object {
  "correctCount": 0,
  "filePath": "D:/Current/image-parsing/__tests__/_data/images-barcode//no-roll.jpg",
  "id": "df8c9ffe-db95-4ccd-a43e-0646905e4992",
  "incorrectCount": 0,
  "isCompiled": false,
  "isRollNoExtracted": false,
  "marks": 0,
  "post": 1,
  "q1": "d",
  "q10": "d",
  "q11": "c",
  "q12": "d",
  "q13": "c",
  "q14": "b",
  "q15": "a",
  "q16": "b",
  "q17": "c",
  "q18": "b",
  "q19": "c",
  "q2": "d",
  "q20": "a",
  "q21": "d",
  "q22": "d",
  "q23": "?",
  "q24": "d",
  "q25": "d",
  "q26": "c",
  "q27": "c",
  "q28": "a",
  "q29": "b",
  "q3": "c",
  "q30": "c",
  "q31": "b",
  "q32": "a",
  "q33": "d",
  "q34": "b",
  "q35": "a",
  "q36": "b",
  "q37": "c",
  "q38": "d",
  "q39": "a",
  "q4": "a",
  "q40": "b",
  "q41": "c",
  "q42": "a",
  "q43": "a",
  "q44": "c",
  "q45": "b",
  "q46": "d",
  "q47": "a",
  "q48": "c",
  "q49": "a",
  "q5": "c",
  "q50": "d",
  "q51": "d",
  "q52": "b",
  "q53": "b",
  "q54": "d",
  "q55": "d",
  "q56": "d",
  "q57": "b",
  "q58": "b",
  "q59": "d",
  "q6": "d",
  "q60": "d",
  "q7": "c",
  "q8": "a",
  "q9": "c",
  "questionPaperType": 1,
  "skippedCount": 0,
  "testCenter": 1,
  "testTime": 1,
  "totalMarks": 0,
  "unattemptedCount": 0,
}
`;

exports[`Result should load from json 1`] = `
Result {
  "answers": Object {
    "q1": Object {
      "value": "a",
    },
    "q2": Object {
      "value": "b",
    },
    "q3": Object {
      "value": "c",
    },
    "q4": Object {
      "value": "d",
    },
    "q5": Object {
      "value": "?",
    },
    "q6": Object {
      "value": "*",
    },
  },
  "correctCount": 0,
  "filePath": "D:\\\\Current\\\\image-parsing\\\\__tests__\\\\_data\\\\images-barcode\\\\10023.jpg",
  "id": "183b5aad-7122-4947-99e9-75c2f73cb076",
  "incorrectCount": 0,
  "isCompiled": false,
  "isRollNoExtracted": true,
  "marks": 0,
  "post": "",
  "questionPaperType": "",
  "rollNo": "10023",
  "skippedCount": 0,
  "testCenter": "",
  "testTime": "",
  "totalMarks": 0,
  "unattemptedCount": 0,
}
`;
